src/service/navigation.js

```diff
@@ -489,3 +527,29 @@ function maybeExpandUrlParams(ampdoc, e) {
     target.setAttribute('href', newHref);
   }
 }
+
+/**
+ * An href transformation applied to an anchor, executed according to its
+ * priority.
+ */
+class LinkRule {
+  /**
+   * @param {function(!Location):!Location} callback
+   * @param {number} priority
+   */
+  constructor(callback, priority) {
+    /** @const @private */
+    this.callback_ = callback;
+    /** @const @private */
+    this.priority_ = priority;
+  }
+
+  /**
+   * Calls the callback for transforming/mutating the location object.
+   * @param {!Location} location
+   * @return {!Location}
+   */
+  transform(location) {
+    return this.callback_(location);
```

Comment:
We talked about using element here. Location seems fine to me, but is there any downside you can thing of @lannka @choumx ?

Reply:
Right, I don't remember the use case for why we needed to use the element here and thought that narrowing it to the scope of the location object would be ok. Lets see what @lannka @choumx say. Thanks for the feedback.

Reply:
I think element is more flexible so that we can
1) read attributes for extra configurations later we might have
2) keep the original URL as an attribute of the element, and can be accessed by any following mutators

Reply:
sounds good. made the change.